src/VarArray.mo

```diff
@@ -107,104 +223,740 @@ module {
+  /// assert VarArray.nextIndexOf<Char>('c', array, 0, Char.equal) == ?0;
+  /// assert VarArray.nextIndexOf<Char>('f', array, 0, Char.equal) == ?2;
+  /// assert VarArray.nextIndexOf<Char>('f', array, 2, Char.equal) == ?2;
+  /// assert VarArray.nextIndexOf<Char>('f', array, 3, Char.equal) == ?3;
+  /// assert VarArray.nextIndexOf<Char>('f', array, 4, Char.equal) == null;
+  /// ```
+  ///
+  /// Runtime: O(array.size())
+  ///
+  /// Space: O(1)
   public func nextIndexOf<T>(element : T, array : [var T], fromInclusive : Nat, equal : (T, T) -> Bool) : ?Nat {
-    todo()
+    var index = fromInclusive;
+    let size = array.size();
+    while (index < size) {
+      if (equal(array[index], element)) {
+        return ?index
+      } else {
+        index += 1
+      }
+    };
+    null
   };
 
+  /// Returns the index of the last `element` in the `array`.
+  ///
+  /// ```motoko include=import
+  /// import Char "mo:base/Char";
+  /// let array = [var 'c', 'o', 'f', 'f', 'e', 'e'];
+  /// assert VarArray.lastIndexOf<Char>('c', array, Char.equal) == ?0;
+  /// assert VarArray.lastIndexOf<Char>('f', array, Char.equal) == ?3;
+  /// assert VarArray.lastIndexOf<Char>('e', array, Char.equal) == ?5;
+  /// assert VarArray.lastIndexOf<Char>('g', array, Char.equal) == null;
+  /// ```
+  ///
+  /// Runtime: O(array.size())
+  ///
+  /// Space: O(1)
   public func lastIndexOf<T>(element : T, array : [var T], equal : (T, T) -> Bool) : ?Nat = prevIndexOf<T>(element, array, array.size(), equal);
 
+  /// Returns the index of the previous occurence of `element` in the `array` starting from the `from` index (exclusive).
+  ///
+  /// ```motoko include=import
+  /// import Char "mo:base/Char";
+  /// let array = [var 'c', 'o', 'f', 'f', 'e', 'e'];
+  /// assert VarArray.prevIndexOf<Char>('c', array, array.size(), Char.equal) == ?0;
+  /// assert VarArray.prevIndexOf<Char>('e', array, array.size(), Char.equal) == ?5;
+  /// assert VarArray.prevIndexOf<Char>('e', array, 5, Char.equal) == ?4;
+  /// assert VarArray.prevIndexOf<Char>('e', array, 4, Char.equal) == null;
+  /// ```
+  ///
+  /// Runtime: O(array.size());
+  /// Space: O(1);
   public func prevIndexOf<T>(element : T, array : [var T], fromExclusive : Nat, equal : (T, T) -> Bool) : ?Nat {
-    todo()
+    var i = fromExclusive;
+    while (i > 0) {
+      i -= 1;
+      if (equal(array[i], element)) {
+        return ?i
+      }
+    };
+    null
   };
 
+  /// Returns an iterator over a slice of the given array.
+  ///
+  /// ```motoko include=import
+  /// let array = [var 1, 2, 3, 4, 5];
+  /// let s = VarArray.slice<Nat>(array, 3, array.size());
+  /// assert s.next() == ?4;
+  /// assert s.next() == ?5;
+  /// assert s.next() == null;
+  ///
+  /// let s = Array.slice<Nat>(array, 0, 0);
+  /// assert s.next() == null;
+  /// ```
+  ///
+  /// Runtime: O(1)
+  ///
+  /// Space: O(1)
   public func slice<T>(array : [var T], fromInclusive : Int, toExclusive : Int) : Iter.Iter<T> {
```

Comment:
I wonder how useful this notion of slice is, since its an iterator.
If this is new, maybe we should postpone its introduction until we decide what a useful notion of slice is.
Maybe we actually want a view on the array with random, not sequential, access to elements.

Reply:
Also why are the bounds Int and not Nat?

Reply:
That's a really good point about potentially wanting to return a view rather than an iterator. Maybe we could rename `slice()` to `range()` due to working in a similar way to ranges over fixed-size numeric types (except where negative numbers are relative to the end of the array)? I updated the PR to show what this would look like, but we could still remove it if you think this might be confusing.